docs/migration-guides/external-providers-primitives-v2.mdx

```diff
@@ -0,0 +1,256 @@
+            3. Instantiate an object which inherits from BasePrimitiveJob 
+                containing pub and runtime information
+            4. Send the job to the execution service of the provider
+            5. Return the data in some format
+        """
+
+        # 1. Define default shots
+        if shots is None:
+            shots = self._default_shots
+
+        # 2. validate SamplerPub format
+        coerced_pubs = [SamplerPub.coerce(pub, shots) for pub in pubs]
+
+        # 3. Instantiate job object, see _run and _run_pub for more details
+        job = PrimitiveJob(self._run, coerced_pubs)
+
+        # 4. Send the job to the execution service of the provider
+        job._submit()
+
+        # 5. Return the data
+        return job
+
+    def _run(self, pubs: Iterable[SamplerPub]) -> PrimitiveResult[SamplerPubResult]:
+        """
+        Iterate over input pubs and call _run_pub.
+        """
+        results = [self._run_pub(pub) for pub in pubs]
+        # return results wrapped in PrimitiveResult
+        return PrimitiveResult(results, metadata={"version": 2})
+
+    def _run_pub(self, pub: SamplerPub) -> SamplerPubResult:
+        """
+        This internal method handles the access to the provider resource.
+        """
+        # pre-processing of the sampling inputs to fit the required format
+        circuit, qargs, meas_info = _preprocess_circuit(pub.circuit)
+        bound_circuits = pub.parameter_values.bind_all(circuit)
+        arrays = {
+            item.creg_name: np.zeros(
+                bound_circuits.shape + (pub.shots, item.num_bytes), dtype=np.uint8
+            )
+            for item in meas_info
+        }
+
+        # iterate over pub elements
+        for index, bound_circuit in enumerate(bound_circuits):
+            # ACCESS PROVIDER RESOURCE HERE
+            # in this case, we are showing an illustrative implementation
+            samples_array = self._hw_resource.sample(bound_circuit)
+
+            # post-processing of the sampling output to fit the required format
+            for item in meas_info:
+                ary = _samples_to_packed_array(samples_array, item.num_bits, item.qreg_indices)
+                arrays[item.creg_name][index] = ary
+        meas = {
+            item.creg_name: BitArray(arrays[item.creg_name], item.num_bits) for item in meas_info
+        }
+        # return results wrapped in SamplerPubResult
+        return SamplerPubResult(
+            DataBin(**meas, shape=pub.shape),
+            metadata={"shots": pub.shots, "circuit_metadata": pub.circuit.metadata},
+        )
+```
+
+This implementation is based on the [`StatevectorSampler` source code](/api/qiskit/qiskit.primitives.StatevectorSampler),
+with some additions to represent the backend/hardware resource access, as the `StatevectorSampler` class is designed for local statevector simulators.
+
+An example of custom `Estimator` implementation for a third-party provider might look like:
```

Comment:
I think it'd be more readable if we have separate sections for Sampler and Estimator examples.